### PR TITLE
refine alsabat test script

### PR DESCRIPTION
**Pull request by keqiaozhang**

### CODE REVIEW

File: test-case/check-alsabat.sh [hunk 1]
```diff
@@ -50,39 +56,51 @@ then
...
+            format_c=S16_LE
+            channel_c=1
+        else
+            format_c=$format
+            channel_c=$channel
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            find /tmp -maxdepth 1 -type f -name "bat.wav.*" -size +0 -exec cp {} "$LOG_ROOT/" \;
```
[the remark this answers is not in the document]
Comment (marc-hb) [reply test-case/check-alsabat.sh:100]: You probably want :
```shell
alsabat ... || {
  find ... || true
  exit 1
}
```
... because of `set -e`. When `find` fails, it fails with an undocumented return value (!) which could cause serious issues like #385

File: test-case/check-alsabat.sh [hunk 2]
```diff
@@ -50,39 +56,51 @@ then
...
+            exit 1
+        }
+        timeout 5 tail --pid=$alsabat_P -f /dev/null || die "Failed to stop alsabat playback"
```
[the remark this answers is not in the document]
Comment (marc-hb) [reply test-case/check-alsabat.sh:103]: The "Failed to stop alsabat" message is misleading, it sounds like `kill` failed. You mean: "alsabat -P failed to stop"

**Review by marc-hb — CHANGES_REQUESTED** [review]
Thanks for a shellcheck-clean PR! Saves a lot of review time.

Any idea about the many check-alsabat-headset-playback.sh failures in https://sof-ci.01.org/softestpr/PR401/build216/devicetest/ ? This doesn't look great in an alsabat PR :-(

I think you can write a much less vague PR title and summary than "refine", just summarize the commit messages.

File: test-case/check-alsabat.sh [hunk 3]
```diff
@@ -50,39 +56,51 @@ then
...
+
+        dlogc "alsabat -P $pcm_p --standalone -n $frames -F $frequency -c $channel -r $rate -f $format"
+        alsabat -P "$pcm_p" --standalone -n "$frames" -F "$frequency" -c "$channel" -r "$rate" -f "$format" & alsabat_P=$!
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
+        sleep 1
+
+        if [ "$codec_mode" != "nocodec" ]; then
+            # USB sound card only supports 1 channel S16_LE format.
+            format_c=S16_LE
+            channel_c=1
```
[the remark this answers is not in the document]
Comment (keqiaozhang) [reply test-case/check-alsabat.sh:95]: nocodec tplgs also support different formats like s16_le/s24_le and s32_le,  not passthrough mode. so we can check each format in the loop.

File: test-case/check-alsabat.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: Good suggestion, I have added a new function in lib.sh to upload wav files.

File: test-case/check-alsabat.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: updated.

File: case-lib/lib.sh [hunk 4]
```diff
@@ -252,6 +252,12 @@ func_lib_get_tplg_path()
     return 0
 }
 
+func_check_nocodec_mode()
+{
+    uname -r |grep "nocodec" >> /dev/null &&
+    echo "nocodec"
+}
+
```
[the remark this answers is not in the document]
Comment (keqiaozhang) [reply case-lib/lib.sh:268]: updated.

File: test-case/check-alsabat.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiulipan) [reply test-case/check-alsabat.sh:95]: Do we real need alsabat for all fmts?

**Review by xiulipan — CHANGES_REQUESTED** [review]

File: test-case/check-alsabat.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (fredoh9) [reply test-case/check-alsabat.sh:95]: Checking all formats available is good idea. but S24_LE is always ignored, isn't it?

File: case-lib/lib.sh [hunk 5]
```diff
@@ -252,6 +252,20 @@ func_lib_get_tplg_path()
...
+    uname -r |grep -q "nocodec"
+}
+
+func_upload_wav_file()
+{
+    local dir=$1
+    local file=$2
+
+    find "$dir" -maxdepth 1 -type f -name "$file" -size +0 -exec cp {} "$LOG_ROOT/" \;
+    return 0
```
Comment (marc-hb) [inline case-lib/lib.sh:266]: The `return 0` will become a no-op when this script is converted to `set -e`. 

I recommend you remove it and you invoke the function like this below:

```shell
func_upload_wav_file "/tmp" "bat.wav.*" || true
```
This gives the caller of the function the choice.

(Cc: @RuiqingX )

File: test-case/check-alsabat.sh [hunk 6]
```diff
@@ -50,39 +56,51 @@ then
...
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            func_upload_wav_file "/tmp" "bat.wav.*"
+            exit 1
+        }
+        wait $alsabatPID || die "alsabat -P failed to stop"
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:103]: It's a bit more complicated, sorry I should have spotted this the first time. To be correct this requires two different error messages. Something like this:
```shell
wait $alsabatPID && ret=$?
if [ $ret = 124 ]; then
   die "alsabat -P failed to stop - we timed it out"
elif [ $ret != 0 ] ; then
   die "alsabat -P failed, returned $ret"
fi
```
Untested code, please double-check in `man timeout` and test by this code by temporarily setting a very short timeout or wrong parameters.

File: test-case/check-alsabat.sh [hunk 7]
```diff
@@ -50,39 +56,51 @@ then
...
+done
+
+exit 0
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:106]: missing newline. There's probably a setting in your editor that you can change to avoid this.

File: test-case/check-alsabat.sh [hunk 8]
```diff
@@ -50,39 +56,51 @@ then
...
+channel=$(func_pipeline_parse_value 0 ch_max)
+rate=$(func_pipeline_parse_value 0 rate)
+fmts=$(func_pipeline_parse_value 0 fmts)
 
 # check the PCMs before alsabat test
 dlogi "check the PCMs before alsabat test"
-[[ $(aplay -Dplug$pcm_p -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
-[[ $(arecord -Dplug$pcm_c -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
+[[ $(aplay -Dplug"$pcm_p" -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
+[[ $(arecord -Dplug"$pcm_c" -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:73]: It's not something you're changing but it looks like this code was wrong. Can you take the opportunity to try to fix it and change it to:
```
aplay -Dplug"$pcm_p" -d 1 /dev/zero -q || die "Failed to play on PCM: $pcm_p"
```

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: test-case/check-alsabat.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (keqiaozhang) [reply test-case/check-alsabat.sh:95]: Yes，S24_4LE is not supported by alsabat

File: test-case/check-alsabat.sh (hunk 8 above)
Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: You are right, die will never be executed.

File: test-case/check-alsabat.sh (hunk 6 above)
Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: Right, if the command times out, and --preserve-status is not set, then exit with status 124.

File: case-lib/lib.sh (hunk 5 above)
Comment (keqiaozhang) [reply case-lib/lib.sh]: fixed

File: test-case/check-alsabat.sh (hunk 6 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh]: > Untested code, please double-check in man timeout and test by this code by temporarily setting a very short timeout or wrong parameters.

With a clearer, Monday morning brain I see the untested code I gave you was wrong: `ret` never gets initialized when `wait` fails because of short circuit evaluation. I'm afraid this means you didn't inject any error locally, didn't test it and never saw these messages printed, this is worrying.

Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: @marc-hb , I did performed the negative tests to verify this part of code, tried with wrong channel count "3" in alsabat command and returned 1.  if alsabat -P does not stop within the specified time, it will return 124. so which scenario you are talking above?

Comment (marc-hb) [reply test-case/check-alsabat.sh]: I mean this:
```shell
ret=0
(exit 5) && ret=$?
echo $ret

=> 0
```

File: test-case/check-alsabat.sh [hunk 9]
```diff
@@ -50,39 +56,56 @@ then
...
+channel=$(func_pipeline_parse_value 0 ch_max)
+rate=$(func_pipeline_parse_value 0 rate)
+fmts=$(func_pipeline_parse_value 0 fmts)
 
 # check the PCMs before alsabat test
 dlogi "check the PCMs before alsabat test"
-[[ $(aplay -Dplug$pcm_p -d 1 /dev/zero -q) ]] && die "Failed to play on PCM: $pcm_p"
-[[ $(arecord -Dplug$pcm_c -d 1 /dev/null -q) ]] && die "Failed to capture on PCM: $pcm_c"
+aplay -Dplug"$pcm_p" -d 1 /dev/zero -q || die "Failed to play on PCM: $pcm_p"
+arecord -Dplug"$pcm_c" -d 1 /dev/null -q || die "Failed to capture on PCM: $pcm_c"
```
Comment (plbossart) [inline test-case/check-alsabat.sh:73]: this looks like an unrelated cleanup?

File: test-case/check-alsabat.sh [hunk 10]
```diff
@@ -50,39 +56,56 @@ then
...
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
+        sleep 1
+
+        if func_nocodec_mode; then
+            format_c=$format
+            channel_c=$channel
+        else
+            # USB sound card only supports 1 channel S16_LE format.
+            format_c=S16_LE
+            channel_c=1
```
Comment (plbossart) [inline test-case/check-alsabat.sh:95]: You guys need to buy better USB cards, this is not acceptable for channel swap tests.

File: test-case/check-alsabat.sh [hunk 11]
```diff
@@ -50,39 +56,56 @@ then
...
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            func_upload_wav_file "/tmp" "bat.wav.*" || true
+            exit 1
+        }
+        wait $alsabatPID && ret=$?
+        if [ $ret == 124 ]; then
```
Comment (plbossart) [inline test-case/check-alsabat.sh:104]: what the hell does 124 mean?

File: test-case/check-alsabat.sh [hunk 12]
```diff
@@ -50,39 +56,56 @@ then
...
-fi
+for format in $fmts
+    do
+        if [ "$format" == "S24_LE" ]; then
+            dlogi "S24_LE is not supported, skip to test this format"
+            continue
+        fi
+
+        dlogc "alsabat -P $pcm_p --standalone -n $frames -F $frequency -c $channel -r $rate -f $format"
+        timeout -k 2 6 alsabat -P "$pcm_p" --standalone -n "$frames" -F "$frequency" -c "$channel" \
```
Comment (plbossart) [inline test-case/check-alsabat.sh:84]: what does 2 6 mean? seriously add comments on what this does.

File: test-case/check-alsabat.sh [hunk 13]
```diff
@@ -50,39 +56,56 @@ then
...
+		-r "$rate" -f "$format" & alsabatPID=$!
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
```
Comment (plbossart) [inline test-case/check-alsabat.sh:86]: avoid

**Review by plbossart — CHANGES_REQUESTED** [review]

File: test-case/check-alsabat.sh (hunk 11 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh:104]: The meaning of 124 is (again) in the `man timeout` page, in a previous review command and much better it is on the very next line: "alsabat -P failed to stop, we timed it out."

File: test-case/check-alsabat.sh (hunk 12 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh:84]: The meanings of 2 and 6 are at the top of the `man timeout` page. I don't think we need to copy/paste into this script the start of the man page of every unusual command but maybe we can have the  `--long` options like this:
```
timeout --kill-after 2 6 alsabat ...
```

File: test-case/check-alsabat.sh (hunk 9 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh:73]: I requested this bug fix, see earlier comments. Agreed it should ideally be in a different commit.

**Review by marc-hb — CHANGES_REQUESTED** [review]
Still concerned about the lack of negative testing mentioned at
https://github.com/thesofproject/sof-test/pull/401#discussion_r502162170

File: test-case/check-alsabat.sh (hunk 6 above)
Comment (keqiaozhang) [reply test-case/check-alsabat.sh]: I think we are overaction here. just put the alsabat -P in background and use wait to check the exit code, if exit code is non-zero， then exit with "alsabat -P failure".

File: test-case/check-alsabat.sh (hunk 9 above)
Comment (keqiaozhang) [reply test-case/check-alsabat.sh:71]: i have separated it as a new commit.

File: test-case/check-alsabat.sh (hunk 10 above)
Comment (xiulipan) [reply test-case/check-alsabat.sh:93]: Same idea here, we are looking for new cards now.

File: case-lib/lib.sh [hunk 14]
```diff
@@ -252,6 +252,19 @@ func_lib_get_tplg_path()
     return 0
 }
 
+func_nocodec_mode()
```
Comment (xiulipan) [inline case-lib/lib.sh:255]: Can we dump the result at first and only use the variable later? Three is no need to check this mode multiple time

File: case-lib/lib.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (xiulipan) [reply case-lib/lib.sh:267]: I do not see any update for this. but I have same idea with this.

**Review by xiulipan — CHANGES_REQUESTED** [review]
Some update is needed. Please check

File: case-lib/lib.sh (hunk 14 above)
Comment (marc-hb) [reply case-lib/lib.sh:255]: > Can we dump the result at first and only use the variable later? 

This would be premature optimization. This command is only forking two additional and small processes per alsabat test, that's negligible.

File: case-lib/lib.sh [hunk 15]
```diff
@@ -252,6 +252,19 @@ func_lib_get_tplg_path()
...
+{
+    uname -r |grep -q "nocodec"
+}
+
+func_upload_wav_file()
+{
+    local dir=$1
+    local file=$2
```
Comment (marc-hb) [inline case-lib/lib.sh:263]: While double quotes are technically not required for assignments it's safer and good role modelling to just always quote.

File: test-case/check-alsabat.sh [hunk 16]
```diff
@@ -48,39 +54,54 @@ then
 	exit 2
 fi
 
+pcmid_p=${pcm_p:-1}
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:57]: Same as above, please double quote for consistency.

File: test-case/check-alsabat.sh [hunk 17]
```diff
@@ -48,39 +54,54 @@ then
...
+            continue
+        fi
+
+        dlogc "alsabat -P $pcm_p --standalone -n $frames -F $frequency -c $channel -r $rate -f $format"
+        alsabat -P "$pcm_p" --standalone -n "$frames" -F "$frequency" -c "$channel" -r "$rate" \
+		-f "$format" & alsabatPID=$!
+        # playback may have low latency, add one second delay to aviod recording zero at beginning.
+        sleep 1
+
+        if func_nocodec_mode; then
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:87]: `if nocodec_mode; then` would read like plain English.

File: test-case/check-alsabat.sh [hunk 18]
```diff
@@ -48,39 +54,54 @@ then
...
+        fi
+
+        dlogc "alsabat -C $pcm_c -F $frequency -f $format_c -c $channel_c -r $rate"
+        alsabat -C "$pcm_c" -F "$frequency" -f "$format_c" -c "$channel_c" -r "$rate" || {
+            func_upload_wav_file "/tmp" "bat.wav.*" || true
+            exit 1
+        }
+        # check the alsabat -P exit code
+        if ! wait $alsabatPID; then
+            die "alsabat -P failure"
```
Comment (marc-hb) [inline test-case/check-alsabat.sh:103]: No need for the `!` negation
```shell
wait $alsabatPID || die "alsabat $alsabatPID failed"
```

File: test-case/check-alsabat.sh (hunk 6 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh]: AFAIK @keqiaozhang didn't recommend to remove the `timeout` command, he only recommended to handle timeout errors exactly like non-timeout errors.

Question for people with `alsabat` experience: can `alsabat` get stuck  when something goes wrong?

**Review by xiulipan — CHANGES_REQUESTED** [review]
@keqiaozhang The PR is almost good, please fix some typo and address the advice to use a variable instead of call `func_nocodec_mode` multiple times

File: test-case/check-alsabat.sh (hunk 10 above)
Comment (marc-hb) [reply test-case/check-alsabat.sh:93]: Internal issue 601 just filed about new USB cards, thanks @singalsu.

Comment (keqiaozhang) [reply test-case/check-alsabat.sh:93]: Thank you all. Let's hold this PR. Now we are trying to find a new USB cards w/ stereo capture support, then we can remove the plug completely.

**Comment by xiulipan** [comment]
@keqiaozhang Ping for update or close.

**Comment by keqiaozhang** [comment]
> @keqiaozhang Ping for update or close.

I just received the new USB card, will try with the new card and refine this patch soon.

**Comment by keqiaozhang** [comment]
Submitted a new one https://github.com/thesofproject/sof-test/pull/554.
